### PR TITLE
Lowers log level of RequestRetriesExhaustedException in the gateway

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

The `RequestRetriesExhaustedException` exception is thrown when requests which should be retried on different partitions (e.g. create workflow or job activation) fail on all partitions __for retry-able errors__ (e.g. not rejections). So the primary case is when back pressure kicks in, which causes an enormous amount of logging (infra team reported our bill went up by 56%). I've lowered it to trace as it can be useful when tracing the gateway to figure out why this happens, and we can now change log levels dynamically.

It should be backported after the PR is approved.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: gateway/src/main/java/io/zeebe/gateway/EndpointManager.java [hunk 1]
```diff
@@ -419,6 +420,12 @@ public static StatusRuntimeException convertThrowable(final Throwable cause) {
       status = Status.NOT_FOUND.augmentDescription(cause.getMessage());
       Loggers.GATEWAY_LOGGER.debug(
           "Expected to handle gRPC request, but request could not be delivered", cause);
+    } else if (cause instanceof RequestRetriesExhaustedException) {
+      // this error occurs when all partitions have exhausted for requests which have no fixed
+      // partitions - it will then also occur when back pressure kicks in, leading to a large burst
+      // of error logs that is, in fact, expected
+      Loggers.GATEWAY_LOGGER.trace(
```
Comment (Zelldon) [inline gateway/src/main/java/io/zeebe/gateway/EndpointManager.java:427]: Should we set the status then to resource exhausted? As we do with the BrokerResourceExhausted exception?

File: gateway/src/main/java/io/zeebe/gateway/impl/broker/RequestRetriesExhaustedException.java [hunk 2]
```diff
@@ -9,7 +9,9 @@
 
 import io.zeebe.gateway.cmd.ClientException;
 
-class RequestRetriesExhaustedException extends ClientException {
+public class RequestRetriesExhaustedException extends ClientException {
+
+  private static final long serialVersionUID = 1493435667474886479L;
```
Comment (Zelldon) [inline gateway/src/main/java/io/zeebe/gateway/impl/broker/RequestRetriesExhaustedException.java:14]: Do you think it is necessary?

File: gateway/src/main/java/io/zeebe/gateway/EndpointManager.java (hunk 1 above)
Comment (Zelldon) [reply gateway/src/main/java/io/zeebe/gateway/EndpointManager.java:427]: Previous it was `unexpected error occured` or what was logged and returned?

**Review by ChrisKujawa — COMMENTED** [review]
Thanks @npepinpe I have some questions before

File: gateway/src/main/java/io/zeebe/gateway/impl/broker/RequestRetriesExhaustedException.java (hunk 2 above)
Comment (npepinpe) [reply gateway/src/main/java/io/zeebe/gateway/impl/broker/RequestRetriesExhaustedException.java:14]: No, sorry, my IDE automatically adds it :see_no_evil:

**Comment by npepinpe** [comment]
Hey, so I thought about it a bit more, and I think I'll approach this a bit differently. Two things I'd like to do:

1. Refactor error mapping logic in a more central location. We can provide a decorated stream observer to handlers such that error mapping is done all in one place.
2. Use a richer error model based on [the following](https://cloud.google.com/apis/design/errors#error_model) (this is taken directly from the gRPC docs, which specifies that most official client/server libraries support this model). This would allow us to add details to specific errors, allowing us to have aggregated errors, such as when we try to send a request over different partitions and may receive different errors.

We could split those up into different issue/PRs as well

**Comment by Zelldon** [comment]
First thing sounds related to https://github.com/zeebe-io/zeebe/issues/5237

**Comment by npepinpe** [comment]
Ah, hadn't seen it. I would bump the priority now imo, seems a good opportunity to do this refactoring before changing the error model. wdyt?

**Comment by Zelldon** [comment]
Sounds good :+1:

**Comment by npepinpe** [comment]
Let's discuss this quickly f2f tomorrow - I have an idea and I want to challenge it before moving ahead with it.

**Comment by npepinpe** [comment]
I updated the PR based on the other one. I tried something which is to add the suppressed exceptions that we add to `RequestRetriesExhaustedException` as error details (so the user can unpack them client side to get more details). I'm not sure about this approach, however. My thought process was that while most of the time, `RequestRetriesExhaustedException` is due to resources exhausted coming from the broker, it's really a more generic error where a request with no explicit partition was tried against all partitions but failed (and not because of a rejection) for all of them - could be one partition was unavailable, one was overloaded, and one had a leader change right as we sent the command. So I wanted to not lose these details - this way the client has more details about what happened, and we reuse the log-level-mapping logic by mapping the suppressed exceptions as well (which weren't logged anywhere before).

**Comment by npepinpe** [comment]
I thought about it a bit more, and I think the right approach is, well, probably to refactor the gateway service such that calls to, say, `activateJobs` would return specific errors and the handler can handle and map them explicitly instead of big `if/else if/else` method. But that means the gateway service (and by this I mean the endpoint manager, the actual business logic) needs to implement an explicit API, and in the gRPC implementation we can explicitly handle these errors and map them. So instead of a big method, we have many small methods which take in specific exception types, and we just call the right one based on which error is returned by the service. But yeah, a bigger refactoring - I would still like to have this PR in as I do want to reduce the logging anyway, and we can try to figure out when's the right time for a gateway refactoring.

**Review by ChrisKujawa — APPROVED** [review]
As you said we should spent at some point more time and love in the gateway and rethink his structure, api and components. 

Looks good for now. Thanks @npepinpe

**Pull request merged**